packages/auth/amplify_auth_cognito_test/test/plugin/fetch_current_device.dart

```diff
@@ -0,0 +1,122 @@
+import 'package:amplify_core/amplify_core.dart';
+import 'package:test/test.dart';
+
+void main() {
+  AmplifyLogger().logLevel = LogLevel.verbose;
+
+  final userPoolKeys = CognitoUserPoolKeys(userPoolConfig);
+  final identityPoolKeys = CognitoIdentityPoolKeys(identityPoolConfig);
+  final testAuthRepo = AmplifyAuthProviderRepository();
+  final mockDevice = DeviceType(deviceKey: deviceKey);
+  final mockDeviceResponse = GetDeviceResponse(device: mockDevice);
+
+  late DeviceMetadataRepository repo;
+  late AmplifyAuthCognitoDart plugin;
+  late CognitoAuthStateMachine stateMachine;
+  late MockSecureStorage secureStorage;
+
+  Future<String?> getDeviceKey() async {
+    final secrets = await repo.get(username);
+    return secrets?.deviceKey;
+  }
+
+  group(
+      'fetchCurrentDevice expected return value and DeviceNotTrackedException',
+      () {
+    setUp(() async {
+      secureStorage = MockSecureStorage();
+      seedStorage(
+        secureStorage,
+        userPoolKeys: userPoolKeys,
+        identityPoolKeys: identityPoolKeys,
+        deviceKeys: CognitoDeviceKeys(userPoolConfig, username),
+      );
+      plugin = AmplifyAuthCognitoDart(
+        secureStorageFactory: (_) => secureStorage,
+      );
+      stateMachine = plugin.stateMachine;
+      await plugin.configure(
+        config: mockConfig,
+        authProviderRepo: testAuthRepo,
+      );
+      final mockIdp = MockCognitoIdentityProviderClient(
+        getDevice: () async => mockDeviceResponse,
+        forgetDevice: () async {},
+      );
+      stateMachine.addInstance<CognitoIdentityProviderClient>(mockIdp);
+      repo = stateMachine.getOrCreate<DeviceMetadataRepository>();
+    });
+
+    test(
+        'should get the current device. current device id should be equal to the local device id',
+        () async {
+      final currentDeviceKey = await getDeviceKey();
+      expect(currentDeviceKey, isNotNull);
+      final currentDevice = await plugin.fetchCurrentDevice();
+      expect(currentDeviceKey, currentDevice.id);
+    });
+
+    test(
+        'should should throw a DeviceNotTrackedException when current device key is null',
+        () async {
+      await plugin.forgetDevice();
+      await expectLater(
+        plugin.fetchCurrentDevice,
+        throwsA(isA<DeviceNotTrackedException>()),
+      );
+    });
+
+    tearDown(() async {
+      await plugin.close();
+    });
+  });
+
+  group('fetchCurrentDevice AWSHttpException', () {
+    setUp(() async {
+      secureStorage = MockSecureStorage();
+      seedStorage(
+        secureStorage,
+        userPoolKeys: userPoolKeys,
+        identityPoolKeys: identityPoolKeys,
+        deviceKeys: CognitoDeviceKeys(userPoolConfig, username),
+      );
+      plugin = AmplifyAuthCognitoDart(
+        secureStorageFactory: (_) => secureStorage,
+      );
+      stateMachine = plugin.stateMachine;
+      await plugin.configure(
+        config: mockConfig,
+        authProviderRepo: testAuthRepo,
+      );
```

Comment:
*nit:* You could move the common setup steps to a single group/setUp fn.

```dart
group('fetchCurrentDevice', () {
  setUp(() async { /* common setup */ });
  group('group 1', () {
    setUp(() async { /* group 1 setup */ });
    // tests
  });
  group('group 2', () {
    setUp(() async { /* group 2 setup */ });
    // tests
  });
});

Reply:
See [response](https://github.com/aws-amplify/amplify-flutter/pull/5074#discussion_r1661553305)

tldr; The top level late declaration gives async children `setUp()`'s access to the asynchronously resolved declarations in its parent `setUp()`'s body. Without it, I can't have a common `setUp()`, and need to decouple and re-declare the mocks.